### PR TITLE
feat: add loading and executing of default client cert source

### PR DESCRIPTION
**Pull request by bshaffer**
Adds logic for detecting and loading the default client certificates for mTLS

Implementation doc: https://google.aip.dev/auth/4114
Related: https://github.com/googleapis/gax-php/pull/325

### CODE REVIEW

File: src/CredentialsLoader.php [hunk 1]
```diff
@@ -247,4 +250,52 @@ public function updateMetadata(
...
+            exec(implode(' ', $cmd), $output, $returnVar);
+
+            if (0 === $returnVar) {
+                return implode(PHP_EOL, $output);
+            }
+            return null;
+        };
+    }
+
+    public static function shouldLoadClientCertSource()
```
Comment (dwsupplee) [inline src/CredentialsLoader.php:282]:
```suggestion
    /**
     * Determines whether or not the default device certificate should be loaded.
     *
     * @return bool
     */
    public static function shouldLoadClientCertSource()
```

File: src/CredentialsLoader.php [hunk 2]
```diff
@@ -247,4 +250,52 @@ public function updateMetadata(
...
+    {
+        return 'true' === getenv(self::MTLS_CERT_ENV_VAR);
```
Comment (dwsupplee) [inline src/CredentialsLoader.php:284]:
```suggestion
        return filter_var(getenv(self::MTLS_CERT_ENV_VAR), FILTER_VALIDATE_BOOLEAN);
```

File: src/CredentialsLoader.php [hunk 3]
```diff
@@ -247,4 +250,52 @@ public function updateMetadata(
 
         return $metadata_copy;
     }
+
+    public static function getDefaultClientCertSource()
+    {
+        if (!$clientCertSourceJson = self::loadDefaultClientCertSourceFile()) {
+            return null;
+        }
+        if (!isset($clientCertSourceJson['cert_provider_command'])) {
```
Comment (dwsupplee) [inline src/CredentialsLoader.php:259]: WDYT about putting this validation in the body of `loadDefaultClientCertSourceFile`?

File: tests/CredentialsLoaderTest.php [hunk 4]
```diff
@@ -29,6 +29,123 @@ public function testUpdateMetadataSkipsWhenAuthenticationisSet()
...
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testActualDefaultClientCertSource()
+    {
+        $clientCertSource = CredentialsLoader::getDefaultClientCertSource();
+        if (is_null($clientCertSource)) {
+            $this->markTestSkipped('No client cert source found');
```
Comment (dwsupplee) [inline tests/CredentialsLoaderTest.php:97]: What are the cases where this wouldn't be found? Should we stub it out?

File: tests/CredentialsLoaderTest.php [hunk 5]
```diff
@@ -29,6 +29,123 @@ public function testUpdateMetadataSkipsWhenAuthenticationisSet()
...
+        $creds = $clientCertSource();
+        $this->assertTrue(is_string($creds));
+        $this->assertContains('-----BEGIN CERTIFICATE-----', $creds);
+        $this->assertContains('-----BEGIN PRIVATE KEY-----', $creds);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testDefaultClientCertSourceInvalidCmdThrowsException()
```
Comment (dwsupplee) [inline tests/CredentialsLoaderTest.php:108]: Not sure if at some point the original intent was to throw an exception, but I do like that approach. If we keep as is, we should update the method here.

```suggestion
    public function testDefaultClientCertSourceInvalidCmdReturnsNull()
```

**Review by dwsupplee — CHANGES_REQUESTED** [review]

File: tests/CredentialsLoaderTest.php (hunk 4 above)
Comment (bshaffer) [reply tests/CredentialsLoaderTest.php:97]: The  client cert source is only found when the corresponding file exists (`~/.secureConnect/context_aware_metadata.json`). This is tested in `testNonExistantDefaultClientCertSource`!

File: tests/CredentialsLoaderTest.php (hunk 5 above)
Comment (bshaffer) [reply tests/CredentialsLoaderTest.php:108]: I agree throwing an exception is better. I can't remember if this was to ensure we don't break customers, or to be consistent with other languages, but I will update it to throw an exception as that's a clearer UX.

File: src/CredentialsLoader.php (hunk 3 above)
Comment (bshaffer) [reply src/CredentialsLoader.php]: done

File: tests/CredentialsLoaderTest.php (hunk 5 above)
Comment (bshaffer) [reply tests/CredentialsLoaderTest.php:110]: Updated to throw an exception if the command errors

**Review by dwsupplee — APPROVED** [review]

**Pull request merged**